File: chat_app/src/lib/api/client/workspace/projects.ts
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// Required Notice: Copyright (c) 2025 AI Chat Team

import { buildQuery } from '../shared';
import type {
  DeleteSuccessResponse,
  ProjectContactLockResponse,
  ProjectContactLinkResponse,
  ProjectPlanOptions,
  ProjectPlanResponse,
  ProjectRequirementWorkItemsOptions,
  ProjectRequirementWorkItemsResponse,
  ProjectRequirementDocumentResponse,
  ProjectRequirementExecuteResponse,
  ProjectRequirementStopResponse,
  ProjectRunEnvironmentResponse,
  ProjectResponse,
  ProjectRunCatalogResponse,
  ProjectRunExecuteResponse,
  ProjectRunStateResponse,
} from '../types';
import type { ApiRequestFn, ContactPaging } from './common';

export const listProjects = (request: ApiRequestFn, userId?: string): Promise<ProjectResponse[]> => {
  const query = buildQuery({ user_id: userId });
  return request<ProjectResponse[]>(`/projects${query}`);
};

export const createProject = (
  request: ApiRequestFn,
  data: { name: string; root_path: string; git_url?: string; description?: string; user_id?: string },
): Promise<ProjectResponse> => {
  return request<ProjectResponse>('/projects', {
    method: 'POST',
    body: JSON.stringify(data),
  });
};

export const updateProject = (
  request: ApiRequestFn,
  id: string,
  data: { name?: string; root_path?: string; git_url?: string; description?: string },
): Promise<ProjectResponse> => {
  return request<ProjectResponse>(`/projects/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  });
};

export const deleteProject = (request: ApiRequestFn, id: string): Promise<DeleteSuccessResponse> => {
  return request<DeleteSuccessResponse>(`/projects/${id}`, {
    method: 'DELETE',
  });
};

export const getProject = (request: ApiRequestFn, id: string): Promise<ProjectResponse> => {
  return request<ProjectResponse>(`/projects/${id}`);
};

export const getProjectPlan = (
  request: ApiRequestFn,
  projectId: string,
  options?: ProjectPlanOptions,
): Promise<ProjectPlanResponse> => {
  const query = buildQuery({
    include_archived: options?.includeArchived,
    include_work_items: options?.includeWorkItems,
  });
  return request<ProjectPlanResponse>(`/projects/${encodeURIComponent(projectId)}/plan${query}`);
};

export const listProjectRequirementWorkItems = (
  request: ApiRequestFn,
  projectId: string,
  requirementId: string,
  options?: ProjectRequirementWorkItemsOptions,
): Promise<ProjectRequirementWorkItemsResponse> => {
  const query = buildQuery({
    include_archived: options?.includeArchived,
    include_dependency_graph: options?.includeDependencyGraph,
  });
  return request<ProjectRequirementWorkItemsResponse>(
    `/projects/${encodeURIComponent(projectId)}/requirements/${encodeURIComponent(requirementId)}/work-items${query}`,
  );
};

export const listProjectRequirementDocuments = (
  request: ApiRequestFn,
  projectId: string,
  requirementId: string,
): Promise<ProjectRequirementDocumentResponse[]> => {
  return request<ProjectRequirementDocumentResponse[]>(
    `/projects/${encodeURIComponent(projectId)}/requirements/${encodeURIComponent(requirementId)}/documents`,
  );
};

export const executeProjectRequirement = (
  request: ApiRequestFn,
  projectId: string,
  requirementId: string,
  data?: { contact_id?: string; include_prerequisite_dependents?: boolean; includePrerequisiteDependents?: boolean },
): Promise<ProjectRequirementExecuteResponse> => {
  return request<ProjectRequirementExecuteResponse>(
    `/projects/${encodeURIComponent(projectId)}/requirements/${encodeURIComponent(requirementId)}/execute`,
    {
      method: 'POST',
      body: JSON.stringify(data || {}),
    },
  );
};

export const stopProjectRequirementExecution = (
  request: ApiRequestFn,
  projectId: string,
  requirementId: string,
  data?: { contact_id?: string },
): Promise<ProjectRequirementStopResponse> => {
  return request<ProjectRequirementStopResponse>(
    `/projects/${encodeURIComponent(projectId)}/requirements/${encodeURIComponent(requirementId)}/stop`,
    {
      method: 'POST',
      body: JSON.stringify(data || {}),
    },
  );
};

export const analyzeProjectRun = (
  request: ApiRequestFn,
  projectId: string,
): Promise<ProjectRunCatalogResponse> => {
  return request<ProjectRunCatalogResponse>(`/projects/${encodeURIComponent(projectId)}/run/analyze`, {
    method: 'POST',
  });
};

export const getProjectRunCatalog = (
  request: ApiRequestFn,
  projectId: string,
): Promise<ProjectRunCatalogResponse> => {
  return request<ProjectRunCatalogResponse>(`/projects/${encodeURIComponent(projectId)}/run/catalog`);
};

export const executeProjectRun = (
  request: ApiRequestFn,
  projectId: string,
  data: {
    target_id?: string;
    cwd?: string;
    command?: string;
    create_if_missing?: boolean;
    terminal_id?: string;
  },
): Promise<ProjectRunExecuteResponse> => {
  return request<ProjectRunExecuteResponse>(`/projects/${encodeURIComponent(projectId)}/run/execute`, {
    method: 'POST',
    body: JSON.stringify(data),
  });
};

export const getProjectRunState = (
  request: ApiRequestFn,
  projectId: string,
): Promise<ProjectRunStateResponse> => {
  return request<ProjectRunStateResponse>(`/projects/${encodeURIComponent(projectId)}/run/state`);
};

export const getProjectRunEnvironment = (
  request: ApiRequestFn,
  projectId: string,
): Promise<ProjectRunEnvironmentResponse> => {
  return request<ProjectRunEnvironmentResponse>(`/projects/${encodeURIComponent(projectId)}/run/environment`);
};

export const updateProjectRunEnvironment = (
  request: ApiRequestFn,
  projectId: string,
  data: {
    selected_toolchains?: Record<string, string>;
    custom_toolchains?: Record<string, { kind?: string; label?: string; path?: string }>;
    env_vars?: Record<string, string>;
    terminal_ui_enabled?: boolean;
  },
): Promise<ProjectRunEnvironmentResponse> => {
  return request<ProjectRunEnvironmentResponse>(`/projects/${encodeURIComponent(projectId)}/run/environment`, {
    method: 'PUT',
    body: JSON.stringify(data),
  });
};

export const setProjectRunDefault = (
  request: ApiRequestFn,
  projectId: string,
  targetId: string,
): Promise<ProjectRunCatalogResponse> => {
  return request<ProjectRunCatalogResponse>(`/projects/${encodeURIComponent(projectId)}/run/default`, {
    method: 'POST',
    body: JSON.stringify({ target_id: targetId }),
  });
};

export const listProjectContacts = (
  request: ApiRequestFn,
  projectId: string,
  paging?: ContactPaging,
): Promise<ProjectContactLinkResponse[]> => {
  const query = buildQuery({
    limit: paging?.limit,
    offset: paging?.offset,
  });
  return request<ProjectContactLinkResponse[]>(`/projects/${encodeURIComponent(projectId)}/contacts${query}`);
};

export const getProjectContactLock = (
  request: ApiRequestFn,
  projectId: string,
): Promise<ProjectContactLockResponse> => {
  return request<ProjectContactLockResponse>(
    `/projects/${encodeURIComponent(projectId)}/contacts/lock`,
  );
};

export const addProjectContact = (
  request: ApiRequestFn,
  projectId: string,
  data: { contact_id: string },
): Promise<ProjectContactLinkResponse> => {
  return request<ProjectContactLinkResponse>(`/projects/${encodeURIComponent(projectId)}/contacts`, {
    method: 'POST',
    body: JSON.stringify(data),
  });
};

export const removeProjectContact = (
  request: ApiRequestFn,
  projectId: string,
  contactId: string,
): Promise<DeleteSuccessResponse> => {
  return request<DeleteSuccessResponse>(
    `/projects/${encodeURIComponent(projectId)}/contacts/${encodeURIComponent(contactId)}`,
    { method: 'DELETE' },
  );
};

File: chatos/backend/src/api/projects.rs
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// Required Notice: Copyright (c) 2025 AI Chat Team

use axum::{
    routing::{delete, get, post},
    Router,
};

mod contact_handlers;
mod contracts;
mod crud_handlers;
pub(crate) mod memory_sync;
mod plan_handlers;
mod requirement_execution;
mod requirement_execution_handlers;
mod run_handlers;
mod session_resolver;

pub(crate) use self::requirement_execution_handlers::{
    cloud_execution_planner_message_is_stale, execution_message_status,
};

pub(crate) async fn repair_stale_cloud_execution_planner_message_for_reconciler(
    message: crate::models::message::Message,
    no_execution_links: bool,
) -> Result<crate::models::message::Message, String> {
    self::requirement_execution_handlers::repair_stale_cloud_execution_planner_message(
        message,
        no_execution_links,
    )
    .await
    .map_err(|err| format!("{err:?}"))
}

pub(crate) async fn reconcile_requirement_planner_owner_context(
    context: serde_json::Value,
) -> Result<(), String> {
    self::requirement_execution_handlers::reconcile_requirement_planner_owner_context(context)
        .await
        .map_err(|error| format!("{error:?}"))
}

use self::contact_handlers::{
    add_project_contact, get_project_contact_lock, list_project_contacts, remove_project_contact,
};
use self::crud_handlers::{delete_project, get_project, list_projects, update_project};
use self::plan_handlers::{
    get_project_plan, list_requirement_documents, list_requirement_work_items,
};
use self::requirement_execution_handlers::{
    confirm_requirement_execution, execute_requirement, get_requirement_execution_plan,
    pause_requirement_execution, rerun_requirement_execution, resume_requirement_execution,
    stop_requirement_execution,
};
use self::run_handlers::{
    analyze_project_run, execute_project_run, get_project_run_catalog, get_project_run_environment,
    get_project_run_state, set_project_run_default, update_project_run_environment,
};
pub fn router() -> Router {
    Router::new()
        .route("/api/projects", get(list_projects))
        .route(
            "/api/projects/{id}",
            get(get_project).put(update_project).delete(delete_project),
        )
        .route("/api/projects/{id}/plan", get(get_project_plan))
        .route(
            "/api/projects/{id}/requirements/{requirement_id}/work-items",
            get(list_requirement_work_items),
        )
        .route(
            "/api/projects/{id}/requirements/{requirement_id}/documents",
            get(list_requirement_documents),
        )
        .route(
            "/api/projects/{id}/requirements/{requirement_id}/execute",
            post(execute_requirement),
        )
        .route(
            "/api/projects/{id}/requirements/{requirement_id}/execution-plan",
            get(get_requirement_execution_plan),
        )
        .route(
            "/api/projects/{id}/requirements/{requirement_id}/confirm-execution",
            post(confirm_requirement_execution),
        )
        .route(
            "/api/projects/{id}/requirements/{requirement_id}/pause",
            post(pause_requirement_execution),
        )
        .route(
            "/api/projects/{id}/requirements/{requirement_id}/resume",
            post(resume_requirement_execution),
        )
        .route(
            "/api/projects/{id}/requirements/{requirement_id}/stop",
            post(stop_requirement_execution),
        )
        .route(
            "/api/projects/{id}/requirements/{requirement_id}/rerun",
            post(rerun_requirement_execution),
        )
        .route(
            "/api/projects/{id}/contacts",
            get(list_project_contacts).post(add_project_contact),
        )
        .route(
            "/api/projects/{id}/contacts/lock",
            get(get_project_contact_lock),
        )
        .route(
            "/api/projects/{id}/contacts/{contact_id}",
            delete(remove_project_contact),
        )
        .route("/api/projects/{id}/run/analyze", post(analyze_project_run))
        .route(
            "/api/projects/{id}/run/catalog",
            get(get_project_run_catalog),
        )
        .route("/api/projects/{id}/run/execute", post(execute_project_run))
        .route("/api/projects/{id}/run/state", get(get_project_run_state))
        .route(
            "/api/projects/{id}/run/default",
            post(set_project_run_default),
        )
        .route(
            "/api/projects/{id}/run/environment",
            get(get_project_run_environment).put(update_project_run_environment),
        )
}
